### PR TITLE
Add getAll test endpoint

### PR DESCRIPTION
**Pull request by Digicat13**
> Created "DTO" project which contains DTOs for entities
Added TestDto, TestQuestionDto, TestAnswerDto.
TestDto contain all information about test, questions of this test and answers for questions.

> Added AutoMapper
Added TestProfile, TestQuestionProfile, TestAnswerProfile in MapperProfiles folder to map Test.cs and TestDto.cs.

> Created "Service" project which contains services interfaces and their implementation
Added folder "Contracts" which contains interfaces for services.
Added TestService to work with Test repository.

> Created TestController
Added GET endpoint and getAll() method using TestService to retrieve information about tests.

> Created ServiceExtension.cs to move services configuration from Startup.cs in separate methods

### CODE REVIEW

File: QuizApp.Services/TestService.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		public TestService(IRepositoryWrapper repository, IMapper mapper)
+		{
+			_repository = repository;
+			_mapper = mapper;
+		}
+
+		public async Task<List<TestDto>> GetAll()
+		{
+			var tests = await _repository.Test.GetAll()
+				.Include(t => t.TestQuestions)
```
Comment (olenarehushleobit) [inline QuizApp.Services/TestService.cs:27]: I would move this code to the repository as it's a part of the logic of just loading data.

File: QuizApp.DAL/Repository/Contracts/IRepository.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
```
Comment (olenarehushleobit) [inline QuizApp.DAL/Repository/Contracts/IRepository.cs:1]: The same issue with branches. Please target the current branch to the branch with repositories, not to the master.

File: QuizApi/ServiceExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using QuizApp.MapperProfiles;
+using QuizApp.DAL.Repository;
+using QuizApp.DAL.Repository.Contracts;
+using QuizApp.Services.Contracts;
+using QuizApp.Services;
+
+namespace QuizApp
+{
+	public static class ServiceExtensions
```
Comment (olenarehushleobit) [inline QuizApi/ServiceExtensions.cs:15]: I would split this class into three different: one for repositories (RepositoryExtension), one for the mapper, and one for services.

File: QuizApi/MapperProfiles/TestQuestionProfile.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace QuizApp.MapperProfiles
+{
+	public class TestQuestionProfile : Profile
+	{
+		public TestQuestionProfile()
+		{
+			CreateMap<TestQuestion, TestQuestionDto>()
+				.ForMember(d => d.Answers, opt => opt.MapFrom(src => src.TestAnswers))
+				.ReverseMap();
```
Comment (olenarehushleobit) [inline QuizApi/MapperProfiles/TestQuestionProfile.cs:17]: Do you really need .`ReverseMap()` here?

File: QuizApi/Controllers/TestController.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+{
+	[Route("api/[controller]")]
+	[ApiController]
+	public class TestController : ControllerBase
+	{
+		private readonly ITestService _testService;
+
+		public TestController(ITestService testService)
+		{
+			_testService = testService;
```
Comment (olenarehushleobit) [inline QuizApi/Controllers/TestController.cs:19]: Please add a check for null reference

File: QuizApp.Services/TestService.cs (hunk 1 above)
Comment (Digicat13) [reply QuizApp.Services/TestService.cs]: Moved Include().ThenInclude() to repository but ToListAsync() is still in the service
Is it ok, or should it also be moved to repository?

File: QuizApi/Controllers/TestController.cs (hunk 5 above)
Comment (Digicat13) [reply QuizApi/Controllers/TestController.cs]: Added check 
_testService ??= testService;
Or there should be testService check and Exception if its null?
_testService = testService ?? throw new ArgumentNullException(nameof(testService));

File: QuizApp.Services/TestService.cs (hunk 1 above)
Comment (olenarehushleobit) [reply QuizApp.Services/TestService.cs]: `ToListAsync()` should also be in the repository

File: QuizApi/Controllers/TestController.cs (hunk 5 above)
Comment (olenarehushleobit) [reply QuizApi/Controllers/TestController.cs]: `_testService = testService ?? throw new ArgumentNullException(nameof(testService));`
this one would be perfect
But let's do this when we will have global exception handling.
You should check the argument for null

File: QuizApi/Controllers/TestController.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+{
+	[Route("api/[controller]")]
+	[ApiController]
+	public class TestController : ControllerBase
+	{
+		private readonly ITestService _testService;
+
+		public TestController(ITestService testService)
+		{
+			_testService ??= testService;
```
Comment (olenarehushleobit) [inline QuizApi/Controllers/TestController.cs:18]: Please check argument `testService` for null, not the readonly property

File: QuizApp.Services/TestService.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace QuizApp.Services
+{
+	public class TestService : ITestService
+	{
+		private readonly IRepositoryWrapper _repository;
+		private readonly IMapper _mapper;
+		public TestService(IRepositoryWrapper repository, IMapper mapper)
+		{
+			_repository ??= repository;
```
Comment (olenarehushleobit) [inline QuizApp.Services/TestService.cs:18]: The same here. Check argument for null

File: QuizApi/Controllers/TestController.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        private readonly ILogger<TestController> _logger;
+
+        public TestController(ITestService testService, ILogger<TestController> logger)
+        {
+            _logger = logger;
+            if (testService == null)
+            {
+                _logger.LogError("Failed to inject TestService into Test controller");
+            }
+            _testService ??= testService;
```
Comment (olenarehushleobit) [inline QuizApi/Controllers/TestController.cs]: you don't need `??` here anymore

**Review by olenarehushleobit — APPROVED** [review]

**Pull request merged**